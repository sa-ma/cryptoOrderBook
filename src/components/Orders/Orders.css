.order-container {
  display: flex;
  width: 100%;
}
.order-container div {
  width: 100%;
}
table {
  border-collapse: collapse;
  width: 100%;
  text-align: center;
}
h2 {
  text-align: center;
}
td,
th {
  padding: 6px;
}
tr:nth-child(even) {
  background-color: #f2f2f2;
}

tr:hover {
  background-color: #ddd;
}
.error {
  text-align: center;
  padding: 2rem;
  font-size: 1.2em;
}
