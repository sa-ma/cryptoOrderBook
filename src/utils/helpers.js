export const subscribeAction = (action, channel) => ({
  event: `bts:${action}`,
  data: {
    channel: `order_book_${channel}`
  }
});

export const getCurrency = (pair) => {
  return pair.toUpperCase().match(/.{1,3}/g);
};
