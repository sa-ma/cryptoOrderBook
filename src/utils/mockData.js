export const initialState = {
  loading: false,
  currencyPairs: [],
  error: {}
};

export const getCurrencyData = {
  data: [
    {
      base_decimals: 8,
      minimum_order: '25.0 USD',
      name: 'LTC/USD',
      counter_decimals: 2,
      trading: 'Enabled',
      url_symbol: 'ltcusd',
      description: 'Litecoin / U.S. dollar'
    }
  ]
};
export const getErrorMessage = 'Request failed with status code 500';
