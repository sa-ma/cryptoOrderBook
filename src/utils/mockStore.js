import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureMockStore([thunk]);

const makeMockStore = (state = {}) => {
  return mockStore({
    ...state
  });
};

export default makeMockStore;
