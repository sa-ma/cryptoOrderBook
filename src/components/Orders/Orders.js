import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { subscribeAction, getCurrency } from '../../utils/helpers';
import PropTypes from 'prop-types';
import './Orders.css';

const Orders = ({ currency }) => {
  const [orders, setOrders] = useState([]);
  const [error, setError] = useState(false);
  let { url_symbol } = currency;
  url_symbol = url_symbol || 'btcusd';

  const currencyArray = getCurrency(url_symbol) || ['BTC', 'USD'];

  useEffect(() => {
    const subscribe = subscribeAction('subscribe', url_symbol);
    const unsubscribe = subscribeAction('unsubscribe', url_symbol);
    const ws = new WebSocket('wss://ws.bitstamp.net');

    const initWebsocket = () => {
      ws.onopen = () => {
        ws.send(JSON.stringify(subscribe));
      };
      ws.onmessage = (event) => {
        const response = JSON.parse(event.data);
        switch (response.event) {
          case 'data':
            setOrders(response.data);
            break;
          case 'bts:request_reconnect':
            initWebsocket();
            break;
          default:
            break;
        }
      };
      ws.onerror = () => {
        setError(true);
      };
      ws.onclose = () => {
        ws.send(JSON.stringify(unsubscribe));
        ws.close();
      };
    };

    initWebsocket();

    return () => {
      ws.close();
    };
  }, [url_symbol]);

  const { bids, asks } = orders;
  const orderRows = (arr) =>
    arr &&
    arr.map((el, index) => (
      <tr key={index}>
        <td> {el[1]} </td>
        <td> {el[0]} </td>
      </tr>
    ));
  const orderHead = (title) => (
    <thead>
      <tr>
        <th colSpan="2">{title}</th>
      </tr>
      <tr>
        <th>Amount ({currencyArray[0]})</th>
        <th>Price ({currencyArray[1]})</th>
      </tr>
    </thead>
  );
  return (
    <div className="order-container">
      {error ? (
        <div className="error"> Error Loading Data</div>
      ) : (
        <>
          {url_symbol ? (
            <div>
              <table>
                {orderHead('Bids')}
                <tbody>{orderRows(bids)}</tbody>
              </table>
            </div>
          ) : (
            <div>Waiting for data ....</div>
          )}
          {url_symbol ? (
            <div>
              <table>
                {orderHead('Asks')}
                <tbody>{orderRows(asks)}</tbody>
              </table>
            </div>
          ) : (
            <div>Waiting for data ....</div>
          )}
        </>
      )}
    </div>
  );
};

Orders.propType = {
  currency: PropTypes.object.isRequired
};
export const OrdersComponent = Orders;
const mapStateToProps = ({ currency }) => ({
  currency: currency.singleCurrency
});
export default connect(mapStateToProps)(Orders);
