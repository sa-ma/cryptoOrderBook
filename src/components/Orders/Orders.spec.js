import React from 'react';
import { mount, shallow } from 'enzyme';
import makeMockStore from '../../utils/mockStore';
import { initialCurrencyState } from '../../utils/mockData';
import { OrdersComponent } from './Orders';

const store = makeMockStore(initialCurrencyState);

describe('<Order/> Component', () => {
  const props = {
    currency: {
      url_symbol: 'ltcusd'
    }
  };
  it('should render the Orders component', () => {
    const component = mount(<OrdersComponent store={store} {...props} />);
    expect(component).toMatchSnapshot();
  });
  it('should successfully display orders and bids', () => {
    const component = shallow(<OrdersComponent store={store} {...props} />);
    expect(component.find('table')).toHaveLength(2);
  });
});
