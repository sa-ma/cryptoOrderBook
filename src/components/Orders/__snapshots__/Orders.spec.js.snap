// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order/> Component should render the Orders component 1`] = `ReactWrapper {}`;
